### PR TITLE
[Dynamo] Fix Sequential nn module with duplicated submodule

### LINKED ISSUE #98852: [torch.compile] `Linear+Relu` * 3 will return different value than without compilation
### 🐛 Describe the bug

[torch.compile] `Linear+Relu` * 3 will return different value than without compilation. It even returns negative value.

```py
import torch
import torch.nn as nn

torch.manual_seed(420)

class Net(nn.Module):

    def __init__(self):
        super(Net, self).__init__()
        self.relu = nn.ReLU()
        self.f = nn.Sequential(nn.Linear(10, 20), self.relu, nn.Linear(20, 30), self.relu, nn.Linear(30, 40), self.relu)

    def forward(self, x):
        x = self.f(x)
        return x
input_tensor = torch.randn(2, 10)

func = Net()
jit_func = torch.compile(func)

print(func(input_tensor))
print(jit_func(input_tensor))
```

```
tensor([[0.0755, 0.1609, 0.0105, 0.0000, 0.0113, 0.0796, 0.0101, 0.0873, 0.1047,
         0.1530, 0.0374, 0.1320, 0.2200, 0.1792, 0.0810, 0.0106, 0.0000, 0.0444,
         0.0422, 0.0000, 0.0000, 0.0172, 0.0000, 0.2407, 0.0920, 0.1108, 0.0000,
         0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0681, 0.1869, 0.2175,
         0.0000, 0.0000, 0.0000, 0.2076],
        [0.1100, 0.1593, 0.0070, 0.0000, 0.0176, 0.0433, 0.0186, 0.0948, 0.0536,
         0.1912, 0.0851, 0.1013, 0.2967, 0.1933, 0.1074, 0.0000, 0.0000, 0.0453,
         0.0394, 0.0000, 0.0000, 0.0167, 0.0000, 0.2231, 0.1349, 0.0448, 0.0587,
         0.0145, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0803, 0.2454, 0.2684,
         0.0000, 0.0000, 0.0000, 0.2026]], grad_fn=<ReluBackward0>)
tensor([[ 0.0766,  0.1797, -0.0009,  0.1803,  0.1252, -0.0972,  0.0340,  0.1121,
          0.1679,  0.2516, -0.1198,  0.1854,  0.1171,  0.1340,  0.0294,  0.0068,
         -0.0920, -0.0025,  0.0129, -0.0945, -0.3049,  0.0083, -0.2495,  0.3413,
          0.0124,  0.0453,  0.0798, -0.2597, -0.1035, -0.1421, -0.0910,  0.0083,
         -0.3073,  0.0124,  0.3064,  0.1881,  0.0574, -0.0936, -0.2189,  0.2819],
        [ 0.1314,  0.1581, -0.0223,  0.1315,  0.1413, -0.1334,  0.0092,  0.1007,
          0.0995,  0.2324, -0.1220,  0.0637,  0.3075,  0.1340, -0.0070, -0.1229,
         -0.1067, -0.0606,  0.0480, -0.0407, -0.4693,  0.0109, -0.2479,  0.2888,
          0.1231,  0.0303,  0.1662, -0.2480, -0.0145, -0.2661, -0.0989,  0.0243,
         -0.3336, -0.0093,  0.3553,  0.2602,  0.0100, -0.0969, -0.2215,  0.2426]],
       grad_fn=<CompiledFunctionBackward>)
```

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230404+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.5-051905-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 510.108.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i9-12900K
CPU family:                      6
Model:                           151
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        2
CPU max MHz:                     6700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6374.40
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       640 KiB (16 instances)
L1i cache:                       768 KiB (16 instances)
L2 cache:                        14 MiB (10 instances)
L3 cache:                        30 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] pytorch-triton==2.1.0+46672772b4
[pip3] torch==2.1.0.dev20230404+cu118
[pip3] torchaudio==2.1.0.dev20230404+cu118
[pip3] torchvision==0.16.0.dev20230404+cu118
[conda] numpy                     1.24.1                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+46672772b4          pypi_0    pypi
[conda] torch                     2.1.0.dev20230404+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230404+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230404+cu118          pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @soumith @msaroufim @wconstab @ngimel @bdhirsh @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### PR DESCRIPTION
**Pull request by yanboliang**
Fixes #98852

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by Chillee — COMMENTED** [review]
Oh... this is not going to be so easy to land. I tried adding something like this a long time ago and it got reverted: https://github.com/pytorch/pytorch/pull/71542

The problem is that folks have subclassed `nn.module` and thus, adding a kwarg to a method is not BC-compatible.

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -2173,7 +2173,7 @@ def children(self) -> Iterator['Module']:
         for name, module in self.named_children():
             yield module
 
-    def named_children(self) -> Iterator[Tuple[str, 'Module']]:
+    def named_children(self, remove_duplicate: bool = True) -> Iterator[Tuple[str, 'Module']]:
```
Comment (albanD) [inline torch/nn/modules/module.py:2176]: Can you update the doc please!

**Review by ngimel — APPROVED** [review]

**Review by albanD — CHANGES_REQUESTED** [review]
As Horace pointed out, some distributed test will most likely fail.
Also please add the proper doc/test in test_nn if you want to land this.

**Comment by yanboliang** [comment]
If the change in ```named_children``` is not easy to land, I don't know if we want to catch any deadline(e.g, 2.0.1?). If we have a tight timeline, I can switch back to use ```mod._modules.items()```. Wdyt?

**Comment by albanD** [comment]
If this is just for Sequential, you can do `for m in module` as well?

**Comment by ngimel** [comment]
It's just for `Sequential`, so `for i,m in enumerate(module)` will work, construct name from `i`

**Comment by ngimel** [comment]
Ok so failing tests indicate that we cannot just replace the name with number if Sequential was constructed with dict, but also it looks like there's no public API to access (duplicated) module names in nn.Sequential?

**Comment by yanboliang** [comment]
Yea, for the following example:
```
class SequentialWithDuplicatedModule2(torch.nn.Module):
    def __init__(self):
        super(SequentialWithDuplicatedModule2, self).__init__()
        self.relu = torch.nn.ReLU()
        self.layer = torch.nn.Sequential(
            collections.OrderedDict(
                [
                    ("linear1", torch.nn.Linear(10, 20)),
                    ("relu1", self.relu),
                    ("linear2", torch.nn.Linear(20, 20)),
                    ("relu2", self.relu),
                    ("linear3", torch.nn.Linear(20, 10)),
                    ("relu3", self.relu),
                ]
            )
        )

    def forward(self, x):
        return self.layer(x)
```
When dynamo registers the module name, we need ```linear1``` rather than ```0```, so ```enumerate(module)``` doesn't work. I'll switch back to ```_modules.items()``` to unblock this issue firstly and we can think how to make ```named_children``` to support ```remove_duplicate``` in a separate PR.

File: torch/_dynamo/variables/nn_module.py [hunk 2]
```diff
@@ -216,7 +216,8 @@ def record_nn_module_stack():
                 # unroll Sequential()
                 assert not kwargs
                 (arg,) = args
-                for child_name, submod in mod.named_children():
+                # TODO: Use named_children when it supports remove_duplicate.
+                for child_name, submod in mod._modules.items():
```
Comment (voznesenskym) [inline torch/_dynamo/variables/nn_module.py:220]: "Why do you need to access that internal attribute instead of using the public APIs?" - @albanD

Comment (yanboliang) [reply torch/_dynamo/variables/nn_module.py:220]: Please read the whole thread, the story behind this decision is clear addressed.

File: torch/_dynamo/variables/nn_module.py [hunk 3]
```diff
@@ -216,7 +216,8 @@ def record_nn_module_stack():
                 # unroll Sequential()
                 assert not kwargs
                 (arg,) = args
-                for child_name, submod in mod.named_children():
+                # TODO: Use named_children when it supports remove_duplicate.
```
Comment (gchanan) [inline torch/_dynamo/variables/nn_module.py:219]: this message is confusing; it sounds like you want remove_duplicate=True, but you actually want remove_duplicate=False, right?

Comment (yanboliang) [reply torch/_dynamo/variables/nn_module.py:219]: Yes, let me make it more clear.

**Pull request closed without merging**